src/python/pants/backend/python/target_types_test.py

```diff
@@ -364,35 +366,90 @@ def test_parse_requirements_file() -> None:
+            *import_rules(),
+            *python_sources.rules(),
             QueryRule(InjectedDependencies, [InjectPythonDistributionDependencies]),
         ],
-        target_types=[PythonDistribution, PexBinary],
+        target_types=[PythonDistribution, PythonRequirementLibrary, PythonLibrary, PexBinary],
         objects={"setup_py": PythonArtifact},
     )
+    rule_runner.add_to_build_file(
+        "",
+        dedent(
+            """\
+            python_requirement_library(
+                name='ansicolors',
+                requirements=['ansicolors'],
+                module_mapping={'ansicolors': ['colors']},
+            )
+            """
+        ),
+    )
+    rule_runner.create_file("project/app.py")
     rule_runner.add_to_build_file(
         "project",
         dedent(
             """\
-            pex_binary(name="my_binary")
+            pex_binary(name="my_binary", entry_point="who_knows.module:main")
+
+            python_library(name="my_library", sources=["app.py"])
+
             python_distribution(
-                name="dist",
+                name="dist-a",
                 provides=setup_py(
-                    name='my-dist'
+                    name='my-dist-a'
                 ).with_binaries({"my_cmd": ":my_binary"})
             )
+
+            python_distribution(
+                name="dist-b",
+                provides=setup_py(
+                    name="my-dist-b"
+                ),
+                entry_points={
+                    "console_scripts":{
+                        "b_cmd": "project.app:main",
+                        "cmd_2": "//project:my_binary",
+                    }
+                },
+            )
             """
         ),
     )
-    tgt = rule_runner.get_target(Address("project", target_name="dist"))
-    injected = rule_runner.request(
-        InjectedDependencies,
-        [InjectPythonDistributionDependencies(tgt[PythonDistributionDependencies])],
+    rule_runner.create_file("who_knows/module.py")
+    rule_runner.add_to_build_file(
+        "who_knows",
+        dedent(
+            """\
+            python_library(name="random_lib", sources=["module.py"])
+            """
+        ),
+    )
+
+    def assert_injected(address: Address, *expected: Address) -> None:
+        tgt = rule_runner.get_target(address)
+        injected = rule_runner.request(
+            InjectedDependencies,
+            [InjectPythonDistributionDependencies(tgt[PythonDistributionDependencies])],
+        )
+        assert injected == InjectedDependencies(expected)
+
+    assert_injected(
+        Address("project", target_name="dist-a"),
+        Address("project", target_name="my_binary"),
+    )
+
+    assert_injected(
+        Address("project", target_name="dist-b"),
+        Address("project", relative_file_path="app.py", target_name="my_library"),
+        Address("who_knows", relative_file_path="module.py", target_name="random_lib"),
```

Comment:
Notice a small difference here, compared to `.with_binaries()` here you get a dependency on the owner of the source file that implements the entry point, whereas for `.with_binaries()` you get a dependency on the `pex_binary` itself.
Not sure if this is cause for any issues, down the road.. ?

Reply:
Oh hm....good eye! You're right that this is a problem. When you have an address to a `pex_binary`, we want a dep on the `pex_binary`, not the `python_library` that has the source code for its entry point.

This will require some changes to your dep inference rule. Right now, you are normalizing the whole `entry_points` field to a set of modules, then doing inference from there. That works when the user specified a module explicitly, but that does not work when they had a target address. Instead, you'll need to differentiate between those two formats when doing inference: if it's a target address, use the same `Get(Addresses, UnparsedAddressInputs)` used for `.with_binaries`, else use the `Get(PythonModuleOwners, PythonModule)` stuff.

> you'll need to differentiate between those two formats when doing inference

Maybe update `ResolvedPythonDistributionEntryPoints` to store two fields: `pex_binary_addresses` and `explicit_modules`, then have a property that merges the two into `all_entry_points`? That allows `setup_py.py` to use `all_entry_points` because it doesn't care about the nuance, but dep inference to use the two fields.